### PR TITLE
Update multiple hypothesis correction and post-process weights matrix

### PR DESCRIPTION
**Pull request by matthewreyna**
This pull request has two updates:

1. The q-values from the Benjamini-Hochberg-Yekutieli procedure were incorrect in same cases, which was can see in the recently added "pancan" example, where some of the p-values were incorrectly larger than the corresponding q-values.  I updated the multiple hypothesis correction code, which returns much more plausible q-values, and the results are consistent with the results from R's `p.adjust` function on the example's p-values and other tests.
2. We can argue that the entries of weight/mutation probability matrix corresponding to gene-patient pairs with the same mutation frequencies should be identical, i.e., entries of the matrix with the same marginals should be the same.  However, because we are sampling from the space of matrices with the same marginals, we do not observe this effect in practice.  I added a simple post-processing procedure that enforces this effect, providing a better approximation of the weight matrix in practice and effectively removing the need for pseudocounts while providing practically no increase in run time.

Each of the results has the following header:
```
#Gene set	WRE (Saddlepoint) P-value	WRE (Saddlepoint) FDR	WRE (Saddlepoint) Runtime	T	Z	t00	t01	t10	t11
```
Here are the top two results from the "pancan" example before the update:
- Across (before update):
```
G1, G2	0.0976427932	0.1302072611	0.0051078796	31	1	68	13	18	1
G3, G4	0.1081409447	0.1302072611	0.0054831505	33	0	67	17	16	0
```
- Between (before update):
```
G3, G4	0.0115878915	0.1293710625	0.0056591034	33	0	67	17	16	0
G1, G2	0.0855275055	0.1293710625	0.0050091743	31	1	68	13	18	1
```
Here are the top two results from the "pancan" example after the update:
- Across (after update):
```
G1, G2	0.0951363459923	1.0	0.00564908981323	31	1	68	13	18	1
G3, G4	0.109830633749	1.0	0.00540399551392	33	0	67	17	16	0
```
- Between (after update):
```
G3, G4	0.0113475661422	1.0	0.00552010536194	33	0	67	17	16	0
G1, G2	0.0839064618439	1.0	0.00510096549988	31	1	68	13	18	1
```
The p-values are nearly the same before and after the updates, but we should see more of a difference with fewer permutations or with infrequently mutated genes and patients.  The q-values are much higher, which is plausible for this example and consistent with the results of R's `p.adjust`.

### CODE REVIEW

File: compute_mutation_probabilities.py [hunk 1]
```diff
@@ -48,6 +48,25 @@ def permute_matrices(edge_list, max_swaps, max_tries, seeds, verbose,
...
+
+    # Find indices corresponding to entries of weight matrix with same marginals
+    marginals_to_indices = defaultdict(list)
+    for i, r_i in enumerate(r):
+        for j, s_j in enumerate(s):
+            marginals_to_indices[(r_i, s_j)].append((i, j))
+
+    # Average weights over entries of weight matrix with same marginals
+    P_mean = np.zeros(np.shape(P))
+    print np.shape(P)
```
Comment (mdml) [inline compute_mutation_probabilities.py]: Logging police! 👮

File: compute_mutation_probabilities.py [hunk 2]
```diff
@@ -119,7 +138,24 @@ def run( args ):
...
+
+        # Post-process weight matrix to assign same weight to entries with same marginals
+        P = postprocess_weight_matrix(P, r, s)
+
+        # Verify the weights again
+        for g, obs in geneToObserved.iteritems():
+            assert( np.abs(P[geneToIndex[g]-1].sum() - obs) < 0.1)
+
+        for p, obs in patientToObserved.iteritems():
+            assert( np.abs(P[:, patientToIndex[p]-1].sum() - obs) < 0.1)
```
Comment (mdml) [inline compute_mutation_probabilities.py]: Why 0.1?

File: compute_mutation_probabilities.py (hunk 1 above)
Comment (mdml) [inline compute_mutation_probabilities.py]: Logging police! 👮

File: compute_mutation_probabilities.py [hunk 3]
```diff
@@ -119,7 +138,24 @@ def run( args ):
...
+
+        # Add pseudocounts to entries with no mutations observed; unlikely except when very few permutations
         P[P == 0] = 1./(2. * args.num_permutations)
```
Comment (mdml) [inline compute_mutation_probabilities.py]: Do you still need this with pseudocounts?

**Review by mdml — APPROVED** [review]
All looks good to me, though I did not rigorously check your `statistics.py` implementations. I did run the pancan example again to make sure it worked for me, and it does.

We should add some testing, ala HotNet2, at some point.

File: compute_mutation_probabilities.py (hunk 2 above)
Comment (matthewreyna) [reply compute_mutation_probabilities.py]: I was hoping that you knew!  I changed it to something more conservative, given that the numerical error will be bounded by the number of floating point operations.

File: compute_mutation_probabilities.py (hunk 3 above)
Comment (matthewreyna) [reply compute_mutation_probabilities.py]: Probably not, but I have not 100% convinced myself yet that there are no edge cases, so I see little harm in leaving it for now.

**Comment by matthewreyna** [comment]
Good questions and comments all.  I pushed minor updates, and I can merge if it looks good.

**Review by mdml — APPROVED** [review]

**Pull request merged**